src/network/live_configuration/update_thread_worker.rs

```diff
@@ -0,0 +1,742 @@
+    current_mode: Arc<Mutex<CurrentMode>>,
+}
+
+impl<T: ServerClient> UpdateThreadWorker<T> {
+    pub(crate) fn new(
+        server_client: T,
+        configuration_id: ConfigurationId,
+        configuration: Arc<Mutex<Option<Configuration>>>,
+        current_mode: Arc<Mutex<CurrentMode>>,
+    ) -> Self {
+        Self {
+            server_client,
+            configuration_id,
+            configuration,
+            current_mode,
+        }
+    }
+
+    /// Executes and _endless_ loop implementing the following behaviour:
+    /// 1. Connects to the websocket
+    /// 2. Retrieves some initial configuration
+    /// 3. Listen to all messages coming from the websocket
+    ///
+    /// This loop will try to keep the connection open until any of these events happen:
+    /// * it receives a termination signal via the `thread_termination_receiver` receiver.
+    /// * it happens any unrecoverable error (see [`UpdateThreadWorker::recoverable_error`])
+    fn run_internal(&self, thread_termination_receiver: Receiver<()>) -> Result<()> {
+        'outer: loop {
+            // Connect websocket, now we are receiving all the update notifications
+            let r = self
+                .server_client
+                .get_configuration_monitoring_websocket(&self.configuration_id);
+            let mut socket = match r {
+                Ok(socket) => socket,
+                Err(e) => {
+                    Self::recoverable_error(e)?;
+                    continue 'outer;
+                }
+            };
+
+            // Get the initial configuration
+            self.update_configuration_from_server_and_current_mode()?;
+
+            'inner: loop {
+                // If the client is gone, we want to exit the loop so the socket is closed on our side, the thread will be terminanted
+                match thread_termination_receiver.try_recv() {
+                    Err(std::sync::mpsc::TryRecvError::Empty) => {}
+                    _ => return Ok(()),
+                }
+
+                // Receive something from the websocket
+                // BUG: If the WS doens't receive data, we are blocked here forever (until the parent process kills this thread).
+                match self.handle_websocket_message(socket)? {
+                    Some(ws) => socket = ws,
+                    None => break 'inner, // Go and create another socket
+                }
+            }
+        }
+    }
+
+    /// Executes [`UpdateThreadWorker::run_internal`] and forwards its result. When this method returns,
+    /// [`UpdateThreadWorker::current_mode`] is set to [`CurrentMode::Defunct`].
+    pub(crate) fn run(&self, thread_termination_receiver: Receiver<()>) -> Result<()> {
+        let result = self.run_internal(thread_termination_receiver);
+        *self.current_mode.lock().unwrap() = CurrentMode::Defunct(result.clone());
+        // TODO: maybe we do not need to return anything here anymore
```

Comment:
Even if we don't _need_ to return anything, it's ok to return the result from `run_internal`, right?